### ISSUE TITLE
Upgrade to graphql-js 14.0

### ISSUE THREAD

**Issue by 19majkel94** [19majkel94/type-graphql#140]
https://github.com/graphql/graphql-js/releases/tag/v14.0.0

- [x] check if `@types/graphql` has been upgraded too 
- [x] wait for fix of result errors typings
- [x] wait for all deps (devDeps) upgrade to v14
- [x] wait for `graphql-query-complexity` case (apollo-server variables)
- [x] upgrade code that use deprecated features (e.g. `introspectionQuery`)
- [x] add `engine` entry to `package.json` + drop support for Node.js 9 in Travis config
- [x] check if all tests pass & fix them
- [x] investigate the need of switching to peer dependency or reexporting `graphql-js` internals

**ETA**: Release 0.15

**Comment by floross** [comment]
>     * [x]  check if `@types/graphql` has been upgraded too

`@types/graphql` has been update to match with the version `14.x.x`
[https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/graphql](url)

**Comment by 19majkel94** [comment]
There is a blocker for now as `graphql-yoga` is not compatible with `graphql-js` v14:
https://github.com/prisma/graphql-yoga/issues/449

PR #152 has been closed as there's plenty of type error to fix to even make the tests pass.

**Comment by kallaspriit** [comment]
I'm in a kind of a deadlock where some of my dependencies require `graphql 14.x.x` but I can't upgrade because `type-graphql` doesn't work with it (`Error: Cannot use GraphQLSchema "[object Object]" from another module or realm.`).

Why is it not possible to upgrade `type-graphql`'s dependencies because `graphql-yoga` has not updated as `graphql-yoga` is only used for examples right, the actual code-base does not depend on it? 

I don't care if the examples break for a while or maybe the examples could be moved to a separate repository/package and just use the older compatible versions?

**Comment by 19majkel94** [comment]
> I'm in a kind of a deadlock where some of my dependencies require graphql 14.x.x but I can't upgrade because type-graphql doesn't work with it

Why you can't just wait with upgrades until all the graphql ecosystem packages upgrade to v14?

> graphql-yoga is only used for examples right

No, it's also used in subscriptions tests.

>  don't care if the examples break for a while or maybe 

But I do care if the tests pass, which is currently not possible due to `graphql-yoga` dependency in tests and typings issue:
https://github.com/DefinitelyTyped/DefinitelyTyped/commit/202ed3a96dabf1121282505fed59de875ecbf077#commitcomment-30611873

**Comment by kallaspriit** [comment]
Ok, I'll try to find a working combination and wait for the ecosystem to catch up with v14. Thanks for the explanation.

**Comment by floross** [comment]
A workaround to use this package right now is to remove the `node_modules` inside `type-graphql` folder.

> rm -r node_modules/type-graphql/node_modules

**Comment by 19majkel94** [comment]
`graphql-yoga` seems abandoned as Apollo Server 2.0 is doing the same job - putting all the subpackages into one nice single API. I've tested it and looks like only the `graphql-subscriptions v1.0` is the blocker for now:
https://github.com/apollographql/apollo-server/pull/1701

**Comment by 19majkel94** [comment]
@kallaspriit 	
For me `apollo-server` feels like a step backward - it's written in typescript but not for typescript consumers in mind:

![image](https://user-images.githubusercontent.com/10618781/46300592-671c0c80-c5a4-11e8-8051-c6a6f40b03a3.png)

Looks like I have to find more time for changing to apollo 😞

**Comment by kallaspriit** [comment]
@19majkel94 are you using `apollo-server` directly or from the `apollo-server-express` etc package? It works fine for me with the express flavor.
![image](https://user-images.githubusercontent.com/277993/46333425-008d0200-c629-11e8-8a30-74bd19007a92.png)

**Comment by 19majkel94** [comment]
@kallaspriit So `ApolloContext` is your custom type? I can't find it in `apollo-server` packages 😕

**Comment by kallaspriit** [comment]
Unfortunately yeah

```typescript
// properties provided to the context function of ApolloServer constructor
export interface ApolloContext {
  req: express.Request;
  [x: string]: any;
}
```

They should change it so that the ApolloServer constructor would accept the user context as a generic  and merge it with the context available by default for express server etc.

**Comment by 19majkel94** [comment]
We have another blockers 😆 	
https://github.com/davidyaha/graphql-redis-subscriptions/issues/127
https://github.com/slicknode/graphql-query-complexity/issues/7#issuecomment-427334091

But I see the light in the end of the tunnel, I hope this week it will be done 😉

**Comment by apiel** [comment]
I did yarn upgrade but I keep getting:
```
[19:36:40] Starting compilation in watch mode...

src/index.ts:53:9 - error TS2322: Type 'import("/home/alex/dev/backend/node_modules/@types/graphql/type/schema").GraphQLSchema' is not assignable to type 'import("/home/alex/dev/backend/node_modules/@types/express-graphql/node_modules/@types/graphql/type/schema").GraphQLSchema'.
  Property 'extensionASTNodes' is missing in type 'GraphQLSchema'.

53         schema,
           ~~~~~~

  node_modules/@types/express-graphql/index.d.ts:32:9
    32         schema: GraphQLSchema;
               ~~~~~~
    The expected type comes from property 'schema' which is declared here on type 'OptionsData'

[19:36:53] Found 1 error. Watching for file changes.
```
How can I solve this properly? Or is this related to this issue?

Right now to solve this I do the following:
`rm -rf node_modules/@types/graphql && cp -r node_modules/@types/express-graphql/node_modules/@types/graphql node_modules/@types/.`

**Comment by 19majkel94** [comment]
Released in `0.15.0`! 🚀
